### PR TITLE
BUG: Fix make_sparse mask generation

### PR DESCRIPTION
**Pull request by Licht-T**
- [x] closes #xxxx
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

This is the part of https://github.com/pandas-dev/pandas/pull/17386.
The mask generation sequence in `make_sparse` treated two data, which have same bits, as same if array is `dtype=object`.
This bug made `SparseArray([False, 0, 100.0, 0.0], fill_value=False, dtype=np.object)` is evaluated as `[False, False, 100.0, False]`

### CODE REVIEW

File: pandas/core/sparse/array.py [hunk 1]
```diff
@@ -789,7 +790,16 @@ def make_sparse(arr, kind='block', fill_value=None):
...
 
-        mask = arr != fill_value
+        if is_object_dtype(arr.dtype):
+            mask = []
+            for e in arr:
+                if type(e) is type(fill_value):
+                    mask.append(e != fill_value)
+                else:
+                    mask.append(True)
+            mask = np.array(mask)
```
Comment (jreback) [inline pandas/core/sparse/array.py]: this is very awkward

you can do this by starting with an array of True
the using .flat to set only the portion you need and do that with a list comprehension

File: pandas/tests/sparse/test_array.py [hunk 2]
```diff
@@ -61,6 +61,12 @@ def test_constructor_object_dtype(self):
         assert arr.dtype == np.object
         assert arr.fill_value == 'A'
 
+        data = [False, 0, 100.0, 0.0]
```
Comment (jreback) [inline pandas/tests/sparse/test_array.py]: add the issues number as a comment

**Review by jreback — CHANGES_REQUESTED** [review]
add a whatsnew entry for this ; use the PR number as issue

**Comment by Licht-T** [comment]
@jreback Thanks for your review.
- Fixed the mask generation method
- Fixed the insufficient array comparison method in test case

File: pandas/core/sparse/array.py [hunk 3]
```diff
@@ -789,7 +790,15 @@ def make_sparse(arr, kind='block', fill_value=None):
         if is_string_dtype(arr):
             arr = arr.astype(object)
 
-        mask = arr != fill_value
+        if is_object_dtype(arr.dtype):
+            mask = np.ones(len(arr), dtype=np.bool)
+            fv_type = type(fill_value)
+
+            itr = (type(x) is fv_type for x in arr)
```
Comment (jreback) [inline pandas/core/sparse/array.py]: this is still quite cumbersome can you simplify

**Comment by Licht-T** [comment]
@jreback The method is now simplified.

File: pandas/core/sparse/array.py [hunk 4]
```diff
@@ -789,7 +790,15 @@ def make_sparse(arr, kind='block', fill_value=None):
         if is_string_dtype(arr):
             arr = arr.astype(object)
 
-        mask = arr != fill_value
+        if is_object_dtype(arr.dtype):
+            mask = np.ones(len(arr), dtype=np.bool)
```
Comment (jreback) [inline pandas/core/sparse/array.py]: ok pls add a comment on what/why this is happening.

no general numpy routine to do this? (I don't think we have a pandas one), but have a look around.

**Review by jreback — CHANGES_REQUESTED** [review]
can you add some asvs for sparse? (if you prefer another issue that's ok too). Need to benchmarks things like this.

**Comment by Licht-T** [comment]
@jreback Added asvs for `SparseArray` constructor, but it does not run due to the bug in `asv_bench/benchmarks/sparse.py`.
I created another PR: #17659.

File: pandas/core/sparse/array.py (hunk 4 above)
Comment (Licht-T) [reply pandas/core/sparse/array.py]: @jreback Added comments. I haven't found yet. I think there is no numpy general method.

**Comment by jreback** [comment]
what do benchmarks show?

**Comment by Licht-T** [comment]
@jreback Here is the results of tests.
This is pretty much slow. Should I reimplement by Cython?

```bash
[asv_bench] asv continuous -f 1.1 pandas-dev/master HEAD -b sparse.sparse_array_constructor                        21:39:21  ☁  fix-make_spase-mask-generation ☂ 𝝙 ⚡ ✭ 𝝙
· Creating environments
· Discovering benchmarks
·· Uninstalling from conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt.
·· Installing into conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt...
· Running 12 total benchmarks (2 commits * 1 environments * 6 benchmarks)
[  0.00%] · For pandas commit hash a48f9572:
[  0.00%] ·· Building for conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt....
[  0.00%] ·· Benchmarking conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[  8.33%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_float64_10percent                                                            39.3±2ms
[ 16.67%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_float64_1percent                                                             7.67±1ms
[ 25.00%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_int64_10percent                                                            39.6±0.4ms
[ 33.33%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_int64_1percent                                                             6.69±0.3ms
[ 41.67%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_object_10percent                                                             556±20ms
[ 50.00%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_object_1percent                                                               486±6ms
[ 50.00%] · For pandas commit hash 5279a172:
[ 50.00%] ·· Building for conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt...................................................................
[ 50.00%] ·· Benchmarking conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 58.33%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_float64_10percent                                                            37.3±3ms
[ 66.67%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_float64_1percent                                                           6.70±0.3ms
[ 75.00%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_int64_10percent                                                              38.0±1ms
[ 83.33%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_int64_1percent                                                             6.16±0.1ms
[ 91.67%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_object_10percent                                                             53.8±2ms
[100.00%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_object_1percent                                                            32.4±0.5ms
       before           after         ratio
     [5279a172]       [a48f9572]
+      32.4±0.5ms          486±6ms    15.01  sparse.sparse_array_constructor.time_sparse_array_constructor_object_1percent
+        53.8±2ms         556±20ms    10.34  sparse.sparse_array_constructor.time_sparse_array_constructor_object_10percent

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
```

**Comment by jreback** [comment]
@Licht-T yes that is slow. Sure if you can do that in cython would be great. its only a single case so shouldn't be much code. not sure how much speedup to get though as you can't really type anything.

**Comment by jreback** [comment]
also, would actually be ok with NOT allowing a not-nan fill_value for ``object`` dtypes as well. Slightly more inflexible but generally using sparse with object dtypes is not very performant anyhow.

**Comment by Licht-T** [comment]
@jreback Here is the Cython result.
```bash
[asv_bench] asv continuous -f 1.1 pandas-dev/master HEAD -b sparse.sparse_array_constructor                        23:04:10  ☁  fix-make_spase-mask-generation-new ☂ ⚡ ✭
· Creating environments
· Discovering benchmarks
·· Uninstalling from conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt.
·· Installing into conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt...
· Running 12 total benchmarks (2 commits * 1 environments * 6 benchmarks)
[  0.00%] · For pandas commit hash ae079f69:
[  0.00%] ·· Building for conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt..................................................................
[  0.00%] ·· Benchmarking conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[  8.33%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_float64_10percent                                                          36.9±0.6ms
[ 16.67%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_float64_1percent                                                           6.91±0.4ms
[ 25.00%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_int64_10percent                                                              38.5±2ms
[ 33.33%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_int64_1percent                                                            5.88±0.05ms
[ 41.67%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_object_10percent                                                             85.9±2ms
[ 50.00%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_object_1percent                                                            52.7±0.6ms
```

**Comment by Licht-T** [comment]
This is rough reimplement and only checking whether this passes tests. It might be a little more faster.

**Comment by Licht-T** [comment]
@jreback Here is the final result. `NaN` and non-`NaN` are almost same.

```bash
[asv_bench] asv continuous -f 1.1 pandas-dev/master HEAD -b sparse.sparse_array_constructor                         0:35:38  ☁  fix-make_spase-mask-generation ☂ 𝝙 ⚡ ✭ 𝝙
· Creating environments
· Discovering benchmarks
·· Uninstalling from conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt.
·· Installing into conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt...
· Running 16 total benchmarks (2 commits * 1 environments * 8 benchmarks)
[  0.00%] · For pandas commit hash 35890726:
[  0.00%] ·· Building for conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt....
[  0.00%] ·· Benchmarking conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[  6.25%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_float64_10percent                                                            40.7±1ms
[ 12.50%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_float64_1percent                                                           7.34±0.1ms
[ 18.75%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_int64_10percent                                                            37.3±0.6ms
[ 25.00%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_int64_1percent                                                            6.34±0.09ms
[ 31.25%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_object_nan_fill_value_10percent                                              78.2±1ms
[ 37.50%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_object_nan_fill_value_1percent                                             44.0±0.6ms
[ 43.75%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_object_non_nan_fill_value_10percent                                          87.0±2ms
[ 50.00%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_object_non_nan_fill_value_1percent                                         51.6±0.5ms
[ 50.00%] · For pandas commit hash 5279a172:
[ 50.00%] ·· Building for conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt....
[ 50.00%] ·· Benchmarking conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 56.25%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_float64_10percent                                                            37.2±2ms
[ 62.50%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_float64_1percent                                                           7.07±0.3ms
[ 68.75%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_int64_10percent                                                              46.8±3ms
[ 75.00%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_int64_1percent                                                            6.66±0.05ms
[ 81.25%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_object_nan_fill_value_10percent                                              82.8±1ms
[ 87.50%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_object_nan_fill_value_1percent                                               46.9±1ms
[ 93.75%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_object_non_nan_fill_value_10percent                                          55.0±2ms
[100.00%] ··· Running sparse.sparse_array_constructor.time_sparse_array_constructor_object_non_nan_fill_value_1percent                                         34.4±0.9ms
       before           after         ratio
     [5279a172]       [35890726]
+        55.0±2ms         87.0±2ms     1.58  sparse.sparse_array_constructor.time_sparse_array_constructor_object_non_nan_fill_value_10percent
+      34.4±0.9ms       51.6±0.5ms     1.50  sparse.sparse_array_constructor.time_sparse_array_constructor_object_non_nan_fill_value_1percent
-        46.8±3ms       37.3±0.6ms     0.80  sparse.sparse_array_constructor.time_sparse_array_constructor_int64_10percent

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
```

**Comment by jreback** [comment]
ok this looks good
going to give a final look in a bit

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm ping on green.

**Comment by jreback** [comment]
thanks @Licht-T keep em coming!

**Pull request merged**